enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/catchup/UpstreamStrategyAddressSupplierTest.java

```diff
@@ -0,0 +1,138 @@
+ * Commons Clause, as found in the associated LICENSE.txt file.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * Neo4j object code can be licensed independently from the source
+ * under separate terms from the AGPL. Inquiries can be directed to:
+ * licensing@neo4j.com
+ *
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.causalclustering.catchup;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.UUID;
+
+import org.neo4j.causalclustering.discovery.TopologyService;
+import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.causalclustering.upstream.UpstreamDatabaseSelectionException;
+import org.neo4j.causalclustering.upstream.UpstreamDatabaseSelectionStrategy;
+import org.neo4j.causalclustering.upstream.UpstreamDatabaseStrategySelector;
+import org.neo4j.helpers.AdvertisedSocketAddress;
+import org.neo4j.logging.NullLogProvider;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class UpstreamStrategyAddressSupplierTest
+{
+    private MemberId self = new MemberId( UUID.randomUUID() );
+    private MemberId other = new MemberId( UUID.randomUUID() );
+    private AdvertisedSocketAddress selfAddr = new AdvertisedSocketAddress( "Self", 123 );
+    private AdvertisedSocketAddress otherAddr = new AdvertisedSocketAddress( "Other", 456 );
+
+    private TopologyService topologyService = mock( TopologyService.class );
+
+    @Before
+    public void setup()
+    {
+        when( topologyService.findCatchupAddress( eq( self ) ) ).thenReturn( Optional.of( selfAddr ) );
+        when( topologyService.findCatchupAddress( eq( other ) ) ).thenReturn( Optional.of( otherAddr ) );
+    }
+
+    @Test
+    public void avoidSelfIfOthersAvailable() throws CatchupAddressResolutionException
+    {
+        // given
+        UpstreamDatabaseStrategySelector upstreamDatabaseStrategySelector =
+                new UpstreamDatabaseStrategySelector( new CountedSelectionStrategy( self, 5 ), Arrays.asList( new CountedSelectionStrategy( other, 2 ) ),
+                        NullLogProvider.getInstance() );
+
+        // and
+        UpstreamStrategyAddressSupplier upstreamStrategyAddressSupplier =
+                new UpstreamStrategyAddressSupplier( upstreamDatabaseStrategySelector, topologyService );
+
+        // when
+        AdvertisedSocketAddress result = upstreamStrategyAddressSupplier.get();
+
+        // then
+        assertEquals( otherAddr, result );
+    }
+
+    @Test
+    public void onlySelfAvailableReturnsSelf() throws CatchupAddressResolutionException
```

Comment:
I think this is wrong. We can never do any catchup request to self

Reply:
Yes, the test is invalid - self checks are now performed per strategy so there is no check in the strategy selector (used in this test). Removed